editoast/src/views/timetable/similar_trains/trains_traffic.rs

```diff
@@ -0,0 +1,307 @@
+    }
+
+    /// Add a train traffic
+    pub fn add_train_traffic(&mut self, train_data: TrainTraffic) -> anyhow::Result<()> {
+        // Only import train that have a valid start date
+        if train_data.start_time > self.traffic_valid_from {
+            // Index the train by its id
+            self.trains_by_id.insert(train_data.id, train_data.clone());
+
+            // Index the train by its speedlimit
+            let trains_for_speedlimit = self
+                .trains_by_speedlimit
+                .entry(train_data.speed_limit_tag)
+                .or_default();
+            trains_for_speedlimit.insert(train_data.id);
+
+            // Index the train by its rollingstock
+            let trains_for_rollingstock = self
+                .trains_by_rollingstock
+                .entry(train_data.rolling_stock.clone())
+                .or_default();
+            trains_for_rollingstock.insert(train_data.id);
+
+            // Index the train by segment
+            for (from, to) in train_data
+                .operational_points
+                .iter()
+                .filter(|op| op.stop)
+                .tuple_windows()
+            {
+                let trains_for_segment = self
+                    .trains_by_segment
+                    .entry(Segment {
+                        from: from.clone(),
+                        to: to.clone(),
+                    })
+                    .or_default();
+                trains_for_segment.insert(train_data.id);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Find in the traffic a composition of trains that match the given train itinerary
+    pub fn find_compatible_trains(
+        &self,
+        rolling_stock: Option<String>,
+        speed_limit_tag: Option<String>,
+        waypoints: Vec<Waypoint>,
+    ) -> Vec<TrainTraffic> {
+        let mut valid_trains = HashSet::<usize>::new();
+
+        // Getting a list of valid train Ids which match the train's specifications
+        let compliant_trains = match (rolling_stock, speed_limit_tag) {
+            (Some(rs), None) => self
+                .trains_by_rollingstock
+                .get(&rs)
+                .cloned()
+                .unwrap_or_default(),
+            (None, Some(speed)) => self
+                .trains_by_speedlimit
+                .get(&speed)
+                .cloned()
+                .unwrap_or_default(),
+            (Some(rs), Some(speed)) => {
+                let rs_trains = self
+                    .trains_by_rollingstock
+                    .get(&rs)
+                    .cloned()
+                    .unwrap_or_default();
+                let speed_trains = self
+                    .trains_by_speedlimit
+                    .get(&speed)
+                    .cloned()
+                    .unwrap_or_default();
+                rs_trains.intersection(&speed_trains).cloned().collect()
+            }
+            _ => self.trains_by_id.keys().cloned().collect(),
```

Comment:
nitpick: maybe the following suggestion helps the reader of the code understanding directly which case is handled here (instead of trying to deduce)... but sure, it's more verbose.


```suggestion
            (None, None) => self.trains_by_id.keys().cloned().collect(),
```